client/fetch_state.go

```
@@ -0,0 +1,169 @@
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/encoding"
+	"github.com/m3db/m3db/serialize"
+	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3db/x/xpool"
+	"github.com/m3db/m3x/ident"
+)
+
+const (
+	maxUint = ^uint(0)
+	maxInt  = int(maxUint >> 1)
+)
+
+var (
+	errFetchStateStillProcessing = errors.New("[invariant violated] fetch state is still processing, unable to create response")
+)
+
+type fetchState struct {
+	sync.Cond
+	sync.Mutex
+	refCounter
+
+	op                   *fetchTaggedOp
+	tagResultAccumulator fetchResultsAccumulator
+	err                  error
+	done                 bool
+
+	pool fetchStatePool
+}
+
+func newFetchState(pool fetchStatePool) *fetchState {
+	f := &fetchState{
+		tagResultAccumulator: &fetchTaggedResultAccumulator{},
+		pool:                 pool,
+	}
+	f.destructorFn = f.close // Set refCounter completion as close
+	f.L = f                  // Set the embedded condition locker to the embedded mutex
+	return f
+}
+
+func (f *fetchState) close() {
+	if f.op != nil {
+		f.op.decRef()
+		f.op = nil
+	}
+	f.err = nil
+	f.done = false
+	f.tagResultAccumulator.Clear()
+
+	if f.pool == nil {
+		return
+	}
+	f.pool.Put(f)
+}
+
+func (f *fetchState) Reset(
+	startTime time.Time,
+	endTime time.Time,
+	op *fetchTaggedOp,
+	topoMap topology.Map,
+	majority int,
+	consistencyLevel topology.ReadConsistencyLevel,
+) {
+	op.incRef() // take a reference to the provided op
+	f.op = op
+	f.tagResultAccumulator.Reset(startTime, endTime, topoMap, majority, consistencyLevel)
+}
+
+func (f *fetchState) completionFn(
+	result interface{},
+	resultErr error,
+) {
+	f.Lock()
+	defer func() {
+		f.Unlock()
+		f.decRef() // release ref held onto by the hostQueue (via op.completionFn)
+	}()
+
+	if f.done {
+		// i.e. we've already failed, no need to continue processing any additional
+		// responses we receive
+		return
+	}
+
+	done, err := f.tagResultAccumulator.Add(result, resultErr)
```

nit: Perhaps tag result accumulator should get the actual casted result, rather than `result interface{}`?  We can return an error earlier here if we know it doesn't cast correctly.

---

sure thing

---

done